### PR TITLE
FEA: fix 1d beam cross section orientation

### LINKED ISSUE #12818: [Problem] Wrong cross section orientation on 1d rectangular beams
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Beam cross section is rotated, when it shouldn't be
MRE

- Create beam with a rectangular (square, should also suffice) cross section in the direction vector (0, 1, 1)
- Observe the rotated cross section after running any analysis

More info can be found at forum thread: https://forum.freecad.org/viewtopic.php?t=85951

### Full version info

```shell
[code]
OS: Windows 10 build 18362
Word size of FreeCAD: 64-bit
Version: 0.21.2.33771 (Git)
Build type: Release
Branch: releases/FreeCAD-0-21
Hash: b9bfa5c5507506e4515816414cd27f4851d00489
Python 3.8.10, Qt 5.15.2, Coin 4.0.1, Vtk 8.2.0, OCC 7.6.3
Locale: English/United States (en_US)
[/code]
```

### Subproject(s) affected?

FEM

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lyphrowny**
closes #12818 

Changes not only the rotation of the cross section, but the order of the 1,2-dimensions for the RECT beam section - now matches with CalculiX docs. This might be a breaking change for some 3rd Parties (if any), but it doesn't break anything in FreeCAD

Tested on Windows

- Example CCX cantilever beam rectangle gives the same result in both the previous and the proposed version
- Checked all directions
all_directions
- Checked ElementRotation1D with different angles
rotated_30deg_cube

.FCStd files are in the forum thread: https://forum.freecad.org/viewtopic.php?t=85951

### CODE REVIEW

**Comment by FEA-eng** [comment]
When I run this modified version of FreeCAD and try to submit the analysis in your test file (011_beam.FCStd), I get the following error:

```
CalculiX finished with error 201
*ERROR in cascade: zero coefficient on the dependent side of an equation
dependent node: 24 direction: 1
```

Same with a simple test file with an inclined rectangular beam. The ones aligned with global axes are working.

**Comment by lyphrowny** [comment]
Thank you for noticing!

### TL;DR

The real issue is with CalculiX, see Dhondtguido/CalculiX/issues/40. It might be better to change the directions of the cross section the way they were, though it's not a panacea

### Long version

Firsty, I reproduced your results on a 011 beam. Then I decided to run a simulation on a all_dirs.FCStd (a bunch of beams, looking at different directions) and strangely there were no errors. 

After that I wrote an example, that generates standalone beams in all directions
The previous implementation [1] failed in 2 directions. Then I tested the current implementation [2] - 5+ fails, and current implementation with FreeCAD-way directions of a cross section [3] - 0 fails. After that I started rotating the beams with the BeamElementRotation1D and FreeCAD directions started failing too.

The CalculiX issue is not fixed in the version, which one can find on the [official site](https://www.dhondt.de/). It is fixed in the original github repo (but I had to revert this commit https://github.com/Dhondtguido/CalculiX/commit/004538d82e2cb1a42a895b3d1ff9b7f75d522974 in order to compile)

As I see it, the best action is to update CalculiX, that's bundled with FreeCAD, the easiest - to have the directions of a cross section the FreeCAD way: unless a rotation is applied, it doesn't fail in any direction (the most common cases)

---
[1] - code can be found here https://github.com/lyphrowny/FreeCAD/tree/all_dirs_examples_generation
[2] - -//- https://github.com/lyphrowny/FreeCAD/tree/check_new_implementation_all_dirs
[3] - -//- https://github.com/lyphrowny/FreeCAD/tree/check_new_implementation_all_dirs_freecad

**Comment by FEA-eng** [comment]
Since it's a bug in the current (2.21) release of CalculiX and the next one likely won't happen anytime soon, perhaps it's better to leave it the FreeCAD way as you say. Using non-stable CalculiX build would be risky. But maybe we could implement a fix for another related limitation and make it possible to use multiple FEM ElementRotation1D objects assigned to different beams. Currently, all beams use the same rotation. What do you think?

**Comment by lyphrowny** [comment]
Yeah, let's leave the directions the FreeCAD way (though I still stand my ground on updating the version of the bundled CalculiX (it's still 2.17 in FreeCAD 0.21.2 on Windows))

We could make it possible to have multiple ElementRotation1D, but it wouldn't solve the issue of 
`*ERROR in cascade: zero coefficient on the dependent side of an equation
dependent node`, but might patch it, i.e. when facing the issue, rotate the beam a bit, repeat until solved

For 1,1,0 beam, which cross section I wanted to be rotated 90 degrees, I managed to rotate it by 73 and 107 degrees, all degrees between 73 and 107 gave me the error in cascade. Is a 17-degrees-off rotation acceptible? :)

**Comment by FEA-eng** [comment]
The CalculiX binary provided with FreeCAD definitely needs to be updated (it seems to be 2.21 in weekly builds already though). It should always be the newest ccx version available when preparing the FreeCAD release (related: #10865) . I’m just unsure about including our own build of ccx created long before the next official release (in which this bug will be fixed) since CalculiX development is in progress and there might be some issues because of that.

Multiple beam rotations are needed mostly for other reasons (more flexibility) but I assumed that they can also help with this issue at least to some extent.

**Comment by lyphrowny** [comment]
I've reverted the change to the directions and tested it locally via `./bin/FreeCADCmd --run-test "TestFemApp"` with no errors
@FEA-eng could you please review?

**Comment by FEA-eng** [comment]
> I've reverted the change to the directions and tested it locally via `./bin/FreeCADCmd --run-test "TestFemApp"` with no errors @FEA-eng could you please review?

I recompiled it. Looks good. It seems to be the best solution for now. But let's keep the issue open or create a new one since there's still some area for improvement once the next version of CalculiX is released.

**Comment by lyphrowny** [comment]
Thank you for the review!
I believe, it should be another issue. #12818 is about wrong cross section orientation and is solved by this PR

**Comment by lyphrowny** [comment]
@FEA-eng please, take a look. I changed the normal for 1-direction from -y to +y as it was in the previous version (only for consistency as -y worked too)

Not all FEM tests were ran via `TestFemApp` unitcase, so it wasn't obvious to me why I was getting errors on builds here, but not on my machine. Is expected behaviour of `TestFemApp`?

**Comment by FEA-eng** [comment]
> @FEA-eng please, take a look. I changed the normal for 1-direction from -y to +y as it was in the previous version (only for consistency as -y worked too)

It's ok for me, works well enough. We have to live with that CalculiX bug for now but at least we know that it should be fixed in the next release.

> Not all FEM tests were ran via `TestFemApp` unitcase, so it wasn't obvious to me why I was getting errors on builds here, but not on my machine. Is expected behaviour of `TestFemApp`?

No idea TBH. I just submit a PR, check the automatic tests and correct what's needed ;-)

**Comment by lyphrowny** [comment]
@chennes could you please start the bulding workflow?

File: src/Mod/Fem/femmesh/meshtools.py [hunk 1]
```diff
@@ -619,87 +619,80 @@ def get_femelement_directions_theshape(femmesh, femelement_table, theshape):
...
+            n = (-1, 0, 0)  # or (0, -1, 0)
+        else:
+            # we want the normal to point downwards
+            n = (x, y, -(x**2 + y**2) / z)
+
+        n = normalize(FreeCAD.Vector(n))
+
+        # sanity check
+        dot = vec.dot(n)
+        if not math.isclose(dot, 0):
```
Comment (lyphrowny) [inline src/Mod/Fem/femmesh/meshtools.py:677]: Evaluated to `False` with `dot` of `1e-18` order of magnitude. Apparently, done the comparison wrong - should've used `abs_tol` parameter (see [stackoverflow](https://stackoverflow.com/a/35325039))
Did a fuzzy testing with vector components in range `[-100_000_000, 100_000_000]` (both floats and ints) and the tests passed with `1e-15` absolute tolerance. Would like to set it to `1e-13` - seems quite acceptable

Comment (lyphrowny) [reply src/Mod/Fem/femmesh/meshtools.py]: Set it to `1e-13`

**Comment by lyphrowny** [comment]
@chennes could you please run the workflow? Should be ok to merge

**Comment by FEA-eng** [comment]
> The CalculiX issue is not fixed in the version, which one can find on the [official site](https://www.dhondt.de/). It is fixed in the original github repo

@lyphrowny A new version of CalculiX (2.22) has just been released. It should have a fix for that bug with beam cross-section orientations. Maybe we can still take it into account and provide a full fix for 1.0?
